lib/src/atom/matcher.rs

```diff
@@ -234,27 +253,40 @@ impl Bindings {
     /// ```
     /// 
     /// TODO: Rename to `add_var_binding` when clients have adopted the new API 
-    pub fn add_var_binding_v2<'a, T1: RefOrMove<VariableAtom>, T2: RefOrMove<Atom>>(mut self, var: T1, value: T2) -> Vec<Bindings> {
-        match self.id_by_var.get(var.as_ref()) {
+    pub fn add_var_binding_v2<T1, T2>(self, var: T1, value: T2) -> Result<Bindings, Atom>
```

Comment:
The `Borrow` trait has a blanket implementation for all types to borrow themselves.  I have confirmed that `borrow().to_owned()` is collapsed into nothing by the compiler when `T` is an owned type.

Unfortunately for our code, it's a moot point because an owned value needs to be cloned (or copied) in order to be used inside a loop.  The first trip through the loop takes ownership and then it's not available to subsequent iterations.  Therefore, in practice, we can either take a reference outside the loop or clone it on every iteration.  Either way it likely compiles down to exactly the same machine code.  The good news is that the compiler is really good at this kind of optimization within a single function or closely coupled group of functions.

Reply:
Yes, it was my first thought that `Borrow` could replace the custom `RefOrMove` interface but I looked at implementation and didn't find the moving. Then I I tried to run small example using `Borrow` and what I see `clone()` is actually called. Please see [rust playground link](https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=201376bf307d8ddc9f07a10f253250b0) and correct me if I am wrong.

I am not sure about which loop are you talking. Do you mean the call of the `match_values()` inside `add_var_binding()`? But there are two others branches of code which don't have a recursion, require moving value and should be used more frequently. I don't insist on keeping `RefOrMove` because it is an experiment and moving semantics is mostly used inside unit tests. Anyway the main usage is done from `match_atoms` which has only references. But may be we can replace type parameters by references and simplify code then.

Reply:
In the case of the playground, by adding special behavior to Clone, the compiler is unable to optimize it away.

However, here is a sandbox where you can see the effect of the compiler's optimizations. 
 https://godbolt.org/z/6966Yjdab

Regarding the loop, I was referring to this line:
```rust
        self.perform_one_to_many_op(|bindings| bindings.add_var_binding_internal(var.borrow(), value.borrow()))
```
In effect, an FnMut closure has all the same limitations as a loop body as far as what it can capture from the outside.

This discussion has made me wonder, why isn't there a more explicit trait to express this.  Even if the compiler makes it unnecessary for performance reasons, the ethos of Rust is to make the code as explicit as possible of the programmer's intentions.  I'll think ask on the forum soon.

Reply:
I added some additional bulk to the struct, to be sure the compiler wasn't taking any shortcuts on account of it being zero-sized.  And also to make a superfluous `clone()` show up better in the asm.  https://godbolt.org/z/Pv95anqEs

The instructions in the `foo_borrowed_to_owned` and `foo_owned` invocations appear to be basically identical.

I'm still wondering why Rust doesn't include a `core` / `std` version of `RefOrMove`.

Reply:
FYI, you're in good company with the `RefOrMove` trait.  https://users.rust-lang.org/t/pattern-for-reformove-arguments/92389

Reply:
Ah ok, I didn't get you were talking about compiler optimizations. It make a perfect sense, thank you for digging into this.

Yes, I overlooked the fact `add_var_binding_internal()` is called from the loop then `RefOrMove` is not applicable. It is interesting if compiler can optimize even this. Probably using loop unrolling. BTW looks like `FnMut` is not really needed here. We can use `Fn` as captured state is not being changed.

Reply:
Yes, I don't have a strong opinion regarding `FnMut`. I agree it is a natural choice for `.for_each()` like functions.

I have no objections on reverting `Borrow` to `RefOrMove`, just afraid doing this is tricky. My main goal was to understand your point and after seeing it I am fine with current solution. But if you would like to do it please feel free doing it.